File: src/extras-kservice/libkserviceaction.cpp
#include <KServiceAction>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVariant>
#include <kserviceaction.h>
#include "libkserviceaction.h"
#include "libkserviceaction.hxx"

KServiceAction* KServiceAction_new() {
    return new KServiceAction();
}

KServiceAction* KServiceAction_new2(const KServiceAction* other) {
    return new KServiceAction(*other);
}

void KServiceAction_OperatorAssign(KServiceAction* self, const KServiceAction* other) {
    self->operator=(*other);
}

void KServiceAction_SetData(KServiceAction* self, const QVariant* userData) {
    self->setData(*userData);
}

QVariant* KServiceAction_Data(const KServiceAction* self) {
    return new QVariant(self->data());
}

libqt_string KServiceAction_Name(const KServiceAction* self) {
    QString _ret = self->name();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KServiceAction_Text(const KServiceAction* self) {
    QString _ret = self->text();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KServiceAction_Icon(const KServiceAction* self) {
    QString _ret = self->icon();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KServiceAction_Exec(const KServiceAction* self) {
    QString _ret = self->exec();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

bool KServiceAction_NoDisplay(const KServiceAction* self) {
    return self->noDisplay();
}

bool KServiceAction_IsSeparator(const KServiceAction* self) {
    return self->isSeparator();
}

void KServiceAction_Delete(KServiceAction* self) {
    delete self;
}

File: src_original/extras-kservice/libkserviceaction.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api.kde.org/kserviceaction.html
pub const kserviceaction = struct {
    /// New constructs a new KServiceAction object.
    ///
    ///
    pub fn New() QtC.KServiceAction {
        return qtc.KServiceAction_new();
    }

    /// New2 constructs a new KServiceAction object.
    ///
    /// ``` other: QtC.KServiceAction ```
    pub fn New2(other: ?*anyopaque) QtC.KServiceAction {
        return qtc.KServiceAction_new2(@ptrCast(other));
    }

    /// [Qt documentation](https://api.kde.org/kserviceaction.html#operator-eq)
    ///
    /// ``` self: QtC.KServiceAction, other: QtC.KServiceAction ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.KServiceAction_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://api.kde.org/kserviceaction.html#setData)
    ///
    /// ``` self: QtC.KServiceAction, userData: QtC.QVariant ```
    pub fn SetData(self: ?*anyopaque, userData: ?*anyopaque) void {
        qtc.KServiceAction_SetData(@ptrCast(self), @ptrCast(userData));
    }

    /// [Qt documentation](https://api.kde.org/kserviceaction.html#data)
    ///
    /// ``` self: QtC.KServiceAction ```
    pub fn Data(self: ?*anyopaque) QtC.QVariant {
        return qtc.KServiceAction_Data(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kserviceaction.html#name)
    ///
    /// ``` self: QtC.KServiceAction, allocator: std.mem.Allocator ```
    pub fn Name(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KServiceAction_Name(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kserviceaction.Name: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kserviceaction.html#text)
    ///
    /// ``` self: QtC.KServiceAction, allocator: std.mem.Allocator ```
    pub fn Text(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KServiceAction_Text(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kserviceaction.Text: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kserviceaction.html#icon)
    ///
    /// ``` self: QtC.KServiceAction, allocator: std.mem.Allocator ```
    pub fn Icon(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KServiceAction_Icon(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kserviceaction.Icon: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kserviceaction.html#exec)
    ///
    /// ``` self: QtC.KServiceAction, allocator: std.mem.Allocator ```
    pub fn Exec(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KServiceAction_Exec(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kserviceaction.Exec: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kserviceaction.html#noDisplay)
    ///
    /// ``` self: QtC.KServiceAction ```
    pub fn NoDisplay(self: ?*anyopaque) bool {
        return qtc.KServiceAction_NoDisplay(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kserviceaction.html#isSeparator)
    ///
    /// ``` self: QtC.KServiceAction ```
    pub fn IsSeparator(self: ?*anyopaque) bool {
        return qtc.KServiceAction_IsSeparator(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kserviceaction.html#dtor.KServiceAction)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.KServiceAction ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.KServiceAction_Delete(@ptrCast(self));
    }
};
